File: src/0144_binary_tree_preorder_traversal/0144_binary_tree_preorder_traversal.py
#!/usr/bin/env python3

from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None) -> None:
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def recusiveHelper(self, root: Optional[TreeNode], values: List[int]) -> None:
        if root is None:
            return

        values.append(root.val)

        self.recusiveHelper(root.left, values)
        self.recusiveHelper(root.right, values)

    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        values = []

        if root is None:
            return values

        values.append(root.val)

        self.recusiveHelper(root.left, values)
        self.recusiveHelper(root.right, values)

        return values

File: solutions/0144_binary_tree_preorder_traversal/0144_binary_tree_preorder_traversal.rs
use std::rc::Rc;
use std::cell::RefCell;

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}
 
impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None
        }
    }
}

impl Solution {
    pub fn preorder_traversal(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
        fn recursive_helper(node: &Option<Rc<RefCell<TreeNode>>>, values: &mut Vec<i32>) {
            if let Some(v) = node {
                let v = v.borrow();

                values.push(v.val);
                recursive_helper(&v.left, values);
                recursive_helper(&v.right, values);
            }
        }

        let mut values: Vec<i32> = Vec::new();

        if let Some(v) = root {
            recursive_helper(&Some(v), &mut values);
        }

        values
    }
}
